modules/packages/arch/metadata.go

```diff
@@ -0,0 +1,330 @@
+// Copyright 2023 The Gitea Authors. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+package arch
+
+import (
+	"archive/tar"
+	"bytes"
+	"compress/gzip"
+	"crypto/md5"
+	"crypto/sha256"
+	"encoding/hex"
+	"errors"
+	"fmt"
+	"io"
+	"os"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/mholt/archiver/v3"
+)
+
+// Metadata for arch package.
+type Metadata struct {
+	Filename          string
+	Name              string
+	Base              string
+	Version           string
+	Description       string
+	CompressedSize    int64
+	CompressedSizeMib string
+	InstalledSize     int64
+	InstalledSizeMib  string
+	MD5               string
+	SHA256            string
+	URL               string
+	BuildDate         int64
+	BuildDateStr      string
+	BaseDomain        string
+	Packager          string
+	Provides          []string
+	License           []string
+	Arch              []string
+	Depends           []string
+	OptDepends        []string
+	MakeDepends       []string
+	CheckDepends      []string
+	Backup            []string
+}
+
+// Function that recieves arch package archive data and returns it's metadata.
+func EjectMetadata(filename, domain string, pkg []byte) (*Metadata, error) {
+	pkgreader := io.LimitReader(bytes.NewReader(pkg), 250000)
+	var buf bytes.Buffer
+	err := archiver.DefaultZstd.Decompress(pkgreader, &buf)
+	if err != nil {
+		if !errors.Is(err, io.ErrUnexpectedEOF) {
+			return nil, err
+		}
+	}
+	splt := strings.Split(buf.String(), "PKGINFO")
+	if len(splt) < 2 {
+		return nil, errors.New("unable to eject .PKGINFO from archive")
+	}
+	raw := splt[1][0:10000]
+	inssize := int64(len(pkg))
+	compsize := ejectInt64(raw, "size")
+	unixbuilddate := ejectInt64(raw, "builddate")
+	return &Metadata{
+		Filename:          filename,
+		Name:              ejectString(raw, "pkgname"),
+		Base:              ejectString(raw, "pkgbase"),
+		Version:           ejectString(raw, "pkgver"),
+		Description:       ejectString(raw, "pkgdesc"),
+		CompressedSize:    inssize,
+		CompressedSizeMib: ByteCountSI(inssize),
+		InstalledSize:     compsize,
+		InstalledSizeMib:  ByteCountSI(compsize),
+		MD5:               md5sum(pkg),
+		SHA256:            sha256sum(pkg),
+		URL:               ejectString(raw, "url"),
+		BuildDate:         unixbuilddate,
+		BuildDateStr:      ReadableTime(unixbuilddate),
+		BaseDomain:        domain,
+		Packager:          ejectString(raw, "packager"),
+		Provides:          ejectStrings(raw, "provides"),
+		License:           ejectStrings(raw, "license"),
+		Arch:              ejectStrings(raw, "arch"),
+		Depends:           ejectStrings(raw, "depend"),
+		OptDepends:        ejectStrings(raw, "optdepend"),
+		MakeDepends:       ejectStrings(raw, "makedepend"),
+		CheckDepends:      ejectStrings(raw, "checkdepend"),
+		Backup:            ejectStrings(raw, "backup"),
```

Comment:
The `eject` functions should just be a switch for the key inside a `Scanner` loop to read the input line by line.